### ISSUE TITLE
consumer stuck during consuming

### ISSUE THREAD

**Issue by abbshr** [Blizzard/node-rdkafka#235]
I use `kafka.KafkaConsumer` API to build a durable etl pipeline. It consume message one by one.

the skeleton is simple:

```coffee
  consumer = new kafka.KafkaConsumer
    "metadata.broker.list": "192.168.5.70:9092"
    "group.id": "node-rdkafka-consumer-1"
    "enable.auto.commit": off

  consumer.connect()

  consumer.on "ready", ->
    consumer.subscribe ["robin-record"]
    consumer.consume 1

  consumer.on "data", (message) ->
    callback message.value.toString()
    consumer.commit message
    consumer.consume 1
```

Sometimes consumer (seems) stucked, because it doesn't consume new message anymore while producer still producing more, thus the lag is getting larger.

- Kafka broker: v0.10.0
- node-rdkafka: v1.0.6

Log from broker is similar to these during consumer consuming:
```
[2017-08-11 15:24:49,574] INFO [GroupCoordinator 1001]: Preparing to restabilize group node-rdkafka-consumer-1 with old generation 46 (kafka.coordinator.GroupCoordinator)
[2017-08-11 15:25:17,031] INFO [GroupCoordinator 1001]: Stabilized group node-rdkafka-consumer-1 generation 47 (kafka.coordinator.GroupCoordinator)
[2017-08-11 15:25:17,062] INFO [GroupCoordinator 1001]: Assignment received from leader for group node-rdkafka-consumer-1 for generation 47 (kafka.coordinator.GroupCoordinator)
[2017-08-11 15:27:53,457] INFO [GroupCoordinator 1001]: Preparing to restabilize group node-rdkafka-consumer-1 with old generation 47 (kafka.coordinator.GroupCoordinator)
[2017-08-11 15:28:21,997] INFO [GroupCoordinator 1001]: Stabilized group node-rdkafka-consumer-1 generation 48 (kafka.coordinator.GroupCoordinator)
[2017-08-11 15:28:22,039] INFO [GroupCoordinator 1001]: Assignment received from leader for group node-rdkafka-consumer-1 for generation 48 (kafka.coordinator.GroupCoordinator)
[2017-08-11 15:30:32,741] INFO [GroupCoordinator 1001]: Preparing to restabilize group node-rdkafka-consumer-1 with old generation 48 (kafka.coordinator.GroupCoordinator)
[2017-08-11 15:30:49,853] INFO [GroupCoordinator 1001]: Stabilized group node-rdkafka-consumer-1 generation 49 (kafka.coordinator.GroupCoordinator)
[2017-08-11 15:31:19,853] INFO [GroupCoordinator 1001]: Preparing to restabilize group node-rdkafka-consumer-1 with old generation 49 (kafka.coordinator.GroupCoordinator)
[2017-08-11 15:31:19,853] INFO [GroupCoordinator 1001]: Group node-rdkafka-consumer-1 with generation 50 is now empty (kafka.coordinator.GroupCoordinator)
```

Didn't find any strange log (as normal).

Then I turn on `debug` options:

```coffee
# "debug": "all"
consumer.on "event.log", (log) ->
  console.log "LOG => "
  console.log log
```

It seems the underlying library didn't stuck, but it didn't fetch any data:

(the current latest offset is 2336891, but the commited offset of this consumer is 2336776)

```
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Topic robin-record [0] MessageSet size 47053, error "Success", MaxOffset 2336891, Ver 2/2' }
LOG =>
{ severity: 7,
  fac: 'CONSUME',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Enqueue 115 messages on robin-record [0] fetch queue (qlen 0, v2)' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Fetch topic robin-record [0] at offset 2336891 (v2)' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Fetch 1/1/1 toppar(s)' }
LOG =>
{ severity: 7,
  fac: 'SEND',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Sent FetchRequest (v1, 71 bytes @ 0, CorrId 10)' }
LOG =>
{ severity: 7,
  fac: 'RECV',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Received FetchResponse (v1, 44 bytes, CorrId 10, rtt 106.95ms)' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Topic robin-record [0] MessageSet size 0, error "Success", MaxOffset 2336891, Ver 2/2' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Fetch backoff for 499ms' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Fetch topic robin-record [0] at offset 2336891 (v2)' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Fetch 1/1/1 toppar(s)' }
LOG =>
{ severity: 7,
  fac: 'SEND',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Sent FetchRequest (v1, 71 bytes @ 0, CorrId 11)' }
LOG =>
{ severity: 7,
  fac: 'RECV',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Received FetchResponse (v1, 44 bytes, CorrId 11, rtt 118.26ms)' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Topic robin-record [0] MessageSet size 0, error "Success", MaxOffset 2336891, Ver 2/2' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Fetch topic robin-record [0] at offset 2336891 (v2)' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Fetch 1/1/1 toppar(s)' }
LOG =>
{ severity: 7,
  fac: 'SEND',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Sent FetchRequest (v1, 71 bytes @ 0, CorrId 12)' }
LOG =>
{ severity: 7,
  fac: 'RECV',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Received FetchResponse (v1, 44 bytes, CorrId 12, rtt 105.60ms)' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Topic robin-record [0] MessageSet size 0, error "Success", MaxOffset 2336891, Ver 2/2' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Fetch topic robin-record [0] at offset 2336891 (v2)' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Fetch 1/1/1 toppar(s)' }
LOG =>
{ severity: 7,
  fac: 'SEND',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Sent FetchRequest (v1, 71 bytes @ 0, CorrId 13)' }
LOG =>
{ severity: 7,
  fac: 'RECV',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Received FetchResponse (v1, 44 bytes, CorrId 13, rtt 107.04ms)' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Topic robin-record [0] MessageSet size 0, error "Success", MaxOffset 2336891, Ver 2/2' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Fetch topic robin-record [0] at offset 2336891 (v2)' }
LOG =>
{ severity: 7,
  fac: 'FETCH',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Fetch 1/1/1 toppar(s)' }
LOG =>
{ severity: 7,
  fac: 'SEND',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Sent FetchRequest (v1, 71 bytes @ 0, CorrId 14)' }
LOG =>
{ severity: 7,
  fac: 'RECV',
  message: '[thrd:192.168.5.70:9092/bootstrap]: 192.168.5.70:9092/1001: Received FetchResponse (v1, 44 bytes, CorrId 14, rtt 106.04ms)' }
```

**Comment by webmakersteve** [comment]
If you only consume during the data callback, if there is an error the error will be silenced, and you will stop consuming.

You need to ensure that consume is getting called independently of whether data was flowing. The consume call can timeout after a given timeout, or when the broker gives a one time error that shouldn't stop execution.

If you use the Consumer stream all of this will be handled for you, and I think it matches your use case better. 

Can you try doing that and see if the issue reproduces itself?

**Comment by abbshr** [comment]
@webmakersteve Yes, u r right. For all important events, I have set triggers on them, but nothing happened during consuming (or blocking).

Besides, I have tried stream API at the beginning. In fact, what I exactly need is **only make sure one message has been ingested (which is async operation) by the external system successfully, should the client commit this offset and proceed ingesting the next message, to make the pipeline durable, or it won't proceed**, but it seems the stream API doesn't strongly guarantee this consuming order because of its concurrency (which I don't need).

**Comment by abbshr** [comment]
Do some trick on stream API, and fixed~

**Comment by prabugp** [comment]
@abbshr what was the trick that you did on the StreamAPI? I'm looking at a similar usecase for my situation as well. Would be great to see how you solved it.

**Comment by rbustami** [comment]
@abbshr @webmakersteve I'm dealing with this same issue when implementing batch consumption and manual commits in v1.0.6.

The round trip time for commits slowly increments until we see a commit failure. The consumers in my group aren't rebalanced and the offset never increments on the partitions where the offset is stuck. 

We ran the same implementation with v2.1.1 and everything works perfectly. I've looked through the release documentation and haven't found anything specifying a fix to this specific issue. 

Any insight as to why we are seeing this problem with v1.0.6 would very helpful.
